File: 1106-parsing-a-boolean-expression/1106-parsing-a-boolean-expression.cpp
class Solution {
public:
    bool parseBoolExpr(string expression) {
        stack<char> st;
        for(auto ch : expression){
            if(ch == ')'){
                bool hasT = false, hasF = false;
                while(st.top() == 't' || st.top() == 'f'){
                    if(st.top() == 'f') hasF = true;
                    if(st.top() == 't') hasT = true;
                    st.pop();
                }
                char op = st.top();
                st.pop();
                if (op == '|') st.push(hasT ? 't' : 'f');
                else if (op == '&') st.push(hasF ? 'f' : 't');
                else if (op == '!') st.push(hasT ? 'f' : 't');
            }
            else if (ch != ',' && ch != '(') st.push(ch);
        }

        return (st.top() == 't') ? true : false;
    }
};

File: 1106-parsing-a-boolean-expression/1106-parsing-a-boolean-expression.java
class Solution {
    public boolean parseBoolExpr(String expression) {
        Stack<Character> st = new Stack<>();
        for(char ch : expression.toCharArray()){
            if(ch == '(' || ch == ',') continue;

            else if(ch == ')'){
                boolean hasF = false, hasT = false;
                while(st.peek() == 't' || st.peek() == 'f'){
                    if(st.peek() == 't') hasT = true;
                    if(st.peek() == 'f') hasF = true;
                    st.pop();
                }
                char op = st.peek();
                st.pop();
                if(op == '!'){
                    if(hasF) st.push('t');
                    else st.push('f');
                }
                else if(op == '&'){
                    if(hasF && hasT) st.push('f');
                    else if(hasF && !hasT) st.push('f');
                    else st.push('t');
                }
                else{
                    if(hasF && hasT) st.push('t');
                    else if(hasF && !hasT) st.push('f');
                    else st.push('t');
                }
            }

            else st.push(ch);
        }

        return (st.peek() == 't') ? true : false;
    }
}